### PR TITLE
Fix clang tidy error and crash

### PR DESCRIPTION
**Pull request by knopp**
Clang tidy checks for https://github.com/flutter/engine/pull/42399 are failing for `formats_mtl.mm` and `context_mtl.mm`. Weirdly this happens even though none of these files where modified in the branch.

For `context_mtl.mm` the fix is straightforward, just a missing `std::move`.
For `formats_mtl.mm` clang-tidy actually crashes on the assignments to `des.depthCompareFunction` and `des.depth_write_enabled`. The crash doesn't happen if the assignment is done through a temporary variable. 

I'm not sure if we should have a workaround for the crash or disable clang-tidy on the file completely (FLUTTER_NOLINT).

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by zanderso — COMMENTED** [review]
Can we hold on this until we see whether the rebase on https://github.com/flutter/engine/pull/42399 helps?

**Comment by zanderso** [comment]
As for the crasher in clang-tidy, @knopp if you can post more details here or in a new GitHub issue, I can try to raise visibility on that with a toolchain team.

**Comment by knopp** [comment]
I don't have much regarding the crash. I can reproduce it locally on main

```
> buildtools/mac-arm64/clang/bin/clang-tidy flutter/impeller/renderer/backend/metal/formats_mtl.mm
[1]    60385 segmentation fault  buildtools/mac-arm64/clang/bin/clang-tidy 
```

<details>
<summary>Stacktrace</summary>
```
  * frame #0: 0x000000010104946a clang-tidy`clang::dataflow::(anonymous namespace)::TransferVisitor::VisitMemberExpr(clang::MemberExpr const*) + 458
    frame #1: 0x000000010104c392 clang-tidy`clang::dataflow::transferCFGBlock(clang::CFGBlock const&, clang::dataflow::AnalysisContext&, std::__2::function<void (clang::CFGElement const&, clang::dataflow::TypeErasedDataflowAnalysisState const&)>) + 8498
    frame #2: 0x000000010104d50d clang-tidy`clang::dataflow::runTypeErasedDataflowAnalysis(clang::dataflow::ControlFlowContext const&, clang::dataflow::TypeErasedDataflowAnalysis&, clang::dataflow::Environment const&, std::__2::function<void (clang::CFGElement const&, clang::dataflow::TypeErasedDataflowAnalysisState const&)>) + 1373
    frame #3: 0x0000000100af7e12 clang-tidy`clang::tidy::bugprone::UncheckedOptionalAccessCheck::check(clang::ast_matchers::MatchFinder::MatchResult const&) + 9938
    frame #4: 0x0000000100720df0 clang-tidy`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::MatchVisitor::visitMatch(clang::ast_matchers::BoundNodes const&) + 352
    frame #5: 0x00000001006e4cbe clang-tidy`clang::ast_matchers::internal::BoundNodesTreeBuilder::visitMatches(clang::ast_matchers::internal::BoundNodesTreeBuilder::Visitor*) + 206
    frame #6: 0x00000001006ed45c clang-tidy`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::matchWithFilter(clang::DynTypedNode const&) + 2028
    frame #7: 0x00000001006ebf10 clang-tidy`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::TraverseDecl(clang::Decl*) + 208
    frame #8: 0x00000001006efd0b clang-tidy`clang::RecursiveASTVisitor<clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor>::TraverseNamespaceDecl(clang::NamespaceDecl*) + 155
    frame #9: 0x00000001006ec16a clang-tidy`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::TraverseDecl(clang::Decl*) + 810
    frame #10: 0x00000001006f915b clang-tidy`clang::RecursiveASTVisitor<clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor>::TraverseTranslationUnitDecl(clang::TranslationUnitDecl*) + 299
    frame #11: 0x00000001006ec7b8 clang-tidy`clang::ast_matchers::internal::(anonymous namespace)::MatchASTVisitor::TraverseDecl(clang::Decl*) + 2424
    frame #12: 0x00000001006ebabd clang-tidy`clang::ast_matchers::MatchFinder::matchAST(clang::ASTContext&) + 749
    frame #13: 0x00000001017b681c clang-tidy`clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) + 44
    frame #14: 0x00000001019fed7a clang-tidy`clang::ParseAST(clang::Sema&, bool, bool) + 13994
    frame #15: 0x00000001017a1985 clang-tidy`clang::FrontendAction::Execute() + 85
    frame #16: 0x00000001016ee894 clang-tidy`clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 1124
    frame #17: 0x0000000100180e26 clang-tidy`clang::tooling::FrontendActionFactory::runInvocation(std::__2::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::__2::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) + 294
    frame #18: 0x000000010076f282 clang-tidy`clang::tidy::runClangTidy(clang::tidy::ClangTidyContext&, clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > >, llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>, bool, bool, llvm::StringRef)::ActionFactory::runInvocation(std::__2::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::__2::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) + 66
    frame #19: 0x000000010076b5d0 clang-tidy`clang::tidy::runClangTidy(clang::tidy::ClangTidyContext&, clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__2::basic_string<char, std::__2::char_traits<char>, std::__2::allocator<char> > >, llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>, bool, bool, llvm::StringRef) + 14400
    frame #20: 0x0000000100901773 clang-tidy`clang::tidy::clangTidyMain(int, char const**) + 27731
    frame #21: 0x0000000204120310 dyld`start + 2432
```
</details>

By experimenting a bit I found out that the crash doesn't happen when the assignment to `des.depthCompareFunction` and `des.depthWriteEnabled` goes through temporary variable.

**Comment by knopp** [comment]
Btw., I think it also crashes my `clangd` as well, so it's not just flutter buildtools. Probably should be reported to LLVM.

File: impeller/renderer/backend/metal/formats_mtl.mm [hunk 1]
```diff
@@ -64,8 +64,13 @@
 
   auto des = [[MTLDepthStencilDescriptor alloc] init];
 
-  des.depthCompareFunction = ToMTLCompareFunction(depth->depth_compare);
-  des.depthWriteEnabled = depth->depth_write_enabled;
+  // These temporary variables are necessary for clang-tidy (Fuchsia LLVM
```
Comment (zanderso) [inline impeller/renderer/backend/metal/formats_mtl.mm:67]: I don't see the crash that this fixes happening on CI. Could we maybe split this out into a separate issue/PR?

Comment (knopp) [reply impeller/renderer/backend/metal/formats_mtl.mm:67]: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8779203752399309425/+/u/test:_lint_host_debug/stdout?format=raw

Screenshot 2023-06-05 at 19 11 32

This is from https://github.com/flutter/engine/pull/42399

I can try to split the PR.

Comment (zanderso) [reply impeller/renderer/backend/metal/formats_mtl.mm:67]: Oh, yeah, sorry. I see it now.

**Review by zanderso — APPROVED** [review]

**Pull request merged**